File: Python/textbook_track/chapter09/ba9j.py
# Reconstruct a String from its Burrows-Wheeler Transform
import sys

from collections import defaultdict


def create_indexed_column(string):
    counts = defaultdict(int)
    char_index = []
    for letter in string:
        count = counts[letter]
        char_index.append((letter, count))
        counts[letter] += 1
    return char_index


def calc_first_column_with_index(transformed_string):
    first_column = create_indexed_column(sorted(transformed_string))
    return {char_index: ix for ix, char_index in enumerate(first_column)}


def calc_inverse_Burrows_Wheeler_transform(transformed_string):
    n = len(transformed_string)
    last_column = create_indexed_column(transformed_string)
    first_column = calc_first_column_with_index(transformed_string)

    position = 0
    string = ['$']
    for _ in range(n-1):
        letter_count = last_column[position]
        string.append(letter_count[0])
        position = first_column[letter_count]
    return ''.join(string[::-1])


def main():
    reader = sys.stdin
    transformed_text = next(reader).rstrip()
    result = calc_inverse_Burrows_Wheeler_transform(transformed_text)
    print(result)


if __name__ == "__main__":
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter09/BA9J.scala
package TextbookTrack.Chapter09

object BA9J {
  import scala.annotation.tailrec

  final case class CharacterIndex(char: Char, number: Int)

  private def createIndexedColumn(string: List[Char]): Vector[CharacterIndex] = {
    @tailrec
    def loop(counts: Map[Char, Int],
             charIndex: List[CharacterIndex],
             xs: List[Char]): Vector[CharacterIndex] = xs match {
      case Nil => charIndex.reverseIterator.toVector
      case x :: xss =>
        val count: Int = counts.getOrElse(x, 0)
        loop(counts.updated(x, count + 1), CharacterIndex(x, count) :: charIndex, xss)
    }

    loop(Map(), Nil, string)
  }

  def calcInverseBurrowsWheelerTransform(transformedText: List[Char]): String = {
    val n: Int = transformedText.length
    val lastColumn: Vector[CharacterIndex] = createIndexedColumn(transformedText)
    val firstColumn: Map[CharacterIndex, Int] = createIndexedColumn(transformedText.sorted).zipWithIndex.toMap

    val (string, _): (List[Char], Int) =
      (0 until (n - 1)).foldLeft((List('$'), 0)){
        case ((acc, position), _) =>
          val letterCount: CharacterIndex = lastColumn(position)
          (letterCount.char :: acc, firstColumn(letterCount))
      }
    string.mkString
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val transformedText: List[Char] = reader.next().toList
    val result: String = calcInverseBurrowsWheelerTransform(transformedText)
    println(result)
  }
}
